### PR TITLE
[None][feat] VisualGen: Add CuTe DSL JIT kernels for FMHA

### PR DESCRIPTION
**Pull request by xrq-phys**
## Summary by CodeRabbit

* **New Features**
  * Added support for runtime-compiled attention kernels, with caching for repeated runs.
  * Expanded attention support to include block-scaled Q/K paths and new FP4-based quantization options.
  * Improved handling of causal, windowed, and optional LSE outputs in attention workflows.

* **Bug Fixes**
  * Tightened validation for attention configurations to better reject unsupported combinations.
  * Removed outdated file-tracking and packaging rules for certain kernel binaries.

## Description

Replace the packaged Blackwell CuTe DSL FMHA binaries with runtime JIT compilation and port the forward kernels and scheduling helpers into VisualGen using public CuTe DSL interfaces.

Add dense QK16PV8 and block-scaled MXFP8/NVFP4 attention paths, including quantized Q/K scale handling and QK smoothing. Extend the VisualGen attention configuration with validated backend-specific recipes while keeping internal kernel controls private.

Update kernel integration, packaging metadata, and unit and performance coverage for the JIT and quantized execution paths.

### Performance (B300, TFLOPS, higher is better)

| Sequence | cubin (QK16PV8) | JIT (QK16PV8) | Difference (QK16PV8) | JIT(MXFP8) | JIT(NVFP4) |
|---:|---:|---:|---:|---:|---:|
| 4,096 | 1,869.2 | 1,814.3 | -2.9% | 1,976.3 | 2,056.2 |
| 32,768 | 1,924.1 | 1,838.0 | -4.5% | 2,206.6 | 2,483.6 |
| 44,160 | 1,908.9 | 1,822.1 | -4.5% | 2,194.1 | 2,458.5 |
| 65,536 | 1,890.2 | 1,855.9 | -1.8% | 2,188.8 | 2,482.4 |
| 98,304 | 1,850.5 | 1,840.0 | -0.6% | 2,184.7 | 2,490.3 |
| 131,072 | 1,843.5 | 1,839.1 | -0.2% | 2,201.0 | 2,495.5 |

Above results use CuTe DSL 4.5.0, as pinned by `requirements.txt`.
Measured from CUPTI device-kernel timestamps collected by PyTorch Profiler/Kineto. Compilation, quantization, allocation, and model execution are excluded.

**if CuTe DSL version was upgraded from 4.5.0 to 4.6.0**

| Sequence | cubin (QK16PV8) | JIT (QK16PV8) | JIT(MXFP8) | JIT(NVFP4) |
|---:|---:|---:|---:|---:|
| 44,160 | 1,909.3 | 1,818.0 | 2,243.9 | 2,540.1 |
| 131,072 | 1,842.6 | 1,839.5 | 2,265.4 | 2,513.3 |

## Test Coverage

- tests/unittest/_torch/visual_gen/test_attention_cute_dsl.py
    - Verifies runtime JIT compilation and compile-cache population on Blackwell GPUs.
    - Compares dense BF16 and QK16PV8 FMHA outputs against PyTorch SDPA across causal/non-causal attention, MHA/GQA, multiple batch sizes, and sequence lengths.
    - Compares block-scaled MXFP8 and NVFP4 outputs against a dequantized BF16 SDPA reference across causal/non-causal and MHA/GQA configurations.
- tests/unittest/_torch/visual_gen/test_visual_gen_args.py
    - Validates supported CUTEDSL recipes for QK16PV8, MXFP8, and NVFP4.
    - Verifies qk_sf_vec handling and rejection of unsupported backend/quantization combinations.
- tests/unittest/_torch/visual_gen/test_attention_integration.py
    - Exercises CUTEDSL through the integrated VisualGen attention module.
    - Checks self-attention, cross-attention, and Wan-shaped attention against reference implementations.
    - Extends CUTEDSL coverage beyond head dimension 128 where supported.
- tests/unittest/_torch/visual_gen/test_attention_perf.py
    - Exercises the JIT CUTEDSL backend in the existing performance matrix, including newly supported head dimensions.

## PR Checklist

Please review the following before submitting your PR:
- PR description clearly explains what and why. If using CodeRabbit's summary, please make sure it makes sense.
- PR Follows [TRT-LLM CODING GUIDELINES](https://github.com/NVIDIA/TensorRT-LLM/blob/main/CODING_GUIDELINES.md) to the best of your knowledge.
- Test cases are provided for new code paths (see [test instructions](https://github.com/NVIDIA/TensorRT-LLM/tree/main/tests#1-how-does-the-ci-work))
- If PR introduces API changes, an appropriate PR label is added - either `api-compatible` or `api-breaking`. For `api-breaking`, include `BREAKING` in the PR title.
- Any new dependencies have been scanned for license and vulnerabilities
- [CODEOWNERS](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/CODEOWNERS) updated if ownership changes
- Documentation updated as needed
- Update [tava architecture diagram](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/tava_architecture_diagram.md) if there is a significant design change in PR.
- The reviewers assigned automatically/manually are appropriate for the PR.

- [x] Please check this after reviewing the above items as appropriate for this PR.

## GitHub Bot Help

To see a list of available CI bot commands, please comment `/bot help`.

### CODE REVIEW

File: tensorrt_llm/_torch/visual_gen/attention_backend/utils.py [hunk 1]
```diff
@@ -49,8 +49,8 @@ def get_visual_gen_attention_backend(
                     Better performance but requires fused QKV
         - "FA4": Flash Attention 4; provides higher speedup on Blackwell GPUs (sm100)
                  Requires flash-attn package with cute interface
-        - "CUTEDSL": CuTe DSL FMHA kernels; uses packaged cubins when present,
-                     otherwise compiles from CuTe DSL source
+        - "CUTEDSL": CuTe DSL FMHA kernels; Best performance on Blackwell GPUs (sm100/sm103)
+                     Compiles from CuTe DSL source
```
Comment (zhenhuaw-me) [inline tensorrt_llm/_torch/visual_gen/attention_backend/utils.py]:
```suggestion
        - "CUTEDSL": CuTe DSL FMHA kernels
```

**Comment by zhenhuaw-me** [comment]
@o-stoner could you take a look at this PR? I am thinking that if any cute DSL infra that we can converge for this PR and https://github.com/NVIDIA/TensorRT-LLM/pull/14280

File: tensorrt_llm/_torch/visual_gen/attention_backend/cute_dsl/fmha.py [hunk 2]
```diff
@@ -15,55 +15,580 @@
...
+        v_qscale = _FP8_E4M3_MAX / v_bshd.float().abs().amax(dim=(0, 1)).clamp(min=1e-3)
+        v_quantized = (v_bshd * v_qscale).to(torch.float8_e4m3fn)
+        return v_quantized, 1.0, v_qscale.reciprocal().contiguous()
+
+    v_qscale = _FP8_E4M3_MAX / v_bshd.abs().amax().clamp(min=1e-3)
+    v_quantized = (v_bshd * v_qscale).to(torch.float8_e4m3fn)
+    return v_quantized, v_qscale.reciprocal(), None
+
+
+def _quantize_blockscaled_one(
```
Comment (zhenhuaw-me) [inline tensorrt_llm/_torch/visual_gen/attention_backend/cute_dsl/fmha.py:506]: I wonder if we already have these utils in the codebase somewhere?

Comment (xrq-phys) [reply tensorrt_llm/_torch/visual_gen/attention_backend/cute_dsl/fmha.py:506]: I was under impression that this utils was only crafted for GEMM?
In fact these one is also a padding/transposed wrapper around GEMM input quantization.

**Review by zhenhuaw-me — COMMENTED** [review]
Can you add some summary about the performance of the kernel?

**Review by zhenhuaw-me — APPROVED** [review]
LGTM thanks!

**Comment by xrq-phys** [comment]
> Can you add some summary about the performance of the kernel?

Thanks. B300 numbers added.

**Review by pengbowang-nv — APPROVED** [review]
Attention part LGTM. Fmha kernel and block sparse kernel are not reviewed line by line as they are from the source of previous cubin.

**Comment by xrq-phys** [comment]
/bot skip --comment "Successful run w/ multi-GPU above. Failure at most recent run is due to unrelated issue."

**Comment by pengbowang-nv** [comment]
Hi @xrq-phys , a kind reminder: I saw most other recent CIs have the same test passed in history panel, maybe you should check why it fails on this branch. And if it is indeed an separate issue maybe we should waive it on main rather than skip the CI.

**Comment by xrq-phys** [comment]
> Hi @xrq-phys , a kind reminder: I saw most other recent CIs have the same test passed in history panel, maybe you should check why it fails on this branch. And if it is indeed an separate issue maybe we should waive it on main rather than skip the CI.

Thanks for the reminder. I saw the A10G error was complaining about missing manifest, presumably due to this VG branch haven't rebased for >1 week. I'll double check multi-GPU runs while waiting for approvals from OSS compliance owners.

**Review by tburt-nv — APPROVED** [review]

**Comment by xrq-phys** [comment]
/bot skip --comment "Re-checked to confirm the only failed case being file outdated file manifest not related to current PR"

**Comment by pengbowang-nv** [comment]
> > Hi @xrq-phys , a kind reminder: I saw most other recent CIs have the same test passed in history panel, maybe you should check why it fails on this branch. And if it is indeed an separate issue maybe we should waive it on main rather than skip the CI.
> 
> Thanks for the reminder. I saw the A10G error was complaining about missing manifest, presumably due to this VG branch haven't rebased for >1 week. I'll double check multi-GPU runs while waiting for approvals from OSS compliance owners.

Hi @xrq-phys . I noticed that the [CI AI analysis](https://pbss.s8k.io/v1/AUTH_svc_tensorrt/sw-tensorrt-ci-analysis/LLM/main/L0_MergeRequest_PR/47690/failure_analysis.html) revealed a different possibility:

> Likely cause: PR #15831 changes telemetry-visible VisualGen argument schema in tensorrt_llm/visual_gen/args.py, including QuantAttentionConfig.qk_dtype, field descriptions, and supported AttentionConfig quantization recipes. The committed golden manifest tensorrt_llm/usage/llm_args_golden_manifest.json was not updated, so the golden-file consistency test fails.

Could you take a look at that and make sure that it is not the change in the PR caused this? Thanks!

**Review by BowenFu — APPROVED** [review]
LGTM — well-guarded and cleanly opt-in. Import failures degrade to `None` (no import-time break), the path is arch-restricted to sm_100a/sm_103a and only activates via `QuantAttentionConfig` on the CUTEDSL backend, so existing BF16/QK16PV8/VSA paths are untouched. Softmax scaling (scale_q·scale_k·sm_scale, log2(e) for exp2; NVFP4/FP8 dequant folded into scales) checks out, and the paths are tested against dequantized SDPA across causal/non-causal + MHA/GQA.

Non-blocking: causal is only tested for square attention (seq_q == seq_kv); if cross-attention causal is ever enabled, add a guard/test for top-left vs bottom-right alignment. Numerical correctness of the ~7k-line CUTLASS-derived port rests entirely on the SDPA-reference unit tests, so those need to actually run on a Blackwell GPU in CI before merge.

**Comment by xrq-phys** [comment]
Thanks so much, Bowen !

Right now I have no causal xattn use case on my immediate list, but will check alignment whenever a use case comes up. Some future possibilities (currently out of scope):
- Fully fused MoT might want top-left alignment in the future
- AR-DiT might need bottom-right in the future.

> Numerical correctness of the ~7k-line CUTLASS-derived port rests entirely on the SDPA-reference unit tests

Thanks for pointing out. Currently VisualGen's attention backends are tested through SDPA-reference + single-block runs + offline video quality checks. IIRC there are some ongoing plans to add e2e video-gen to the tests but still iterating on the time/coverage trade-off.

**Comment by xrq-phys** [comment]
> > Likely cause: PR #15831 changes telemetry-visible VisualGen argument schema in tensorrt_llm/visual_gen/args.py, including QuantAttentionConfig.qk_dtype, field descriptions, and supported AttentionConfig quantization recipes. The committed golden manifest tensorrt_llm/usage/llm_args_golden_manifest.json was not updated, so the golden-file consistency test fails.
> 
> Could you take a look at that and make sure that it is not the change in the PR caused this? Thanks!

Thanks for the comment, @pengbowang-nv !

I tried to rerun `python3 scripts/generate_llm_args_golden_manifest.py` but the golden file didn't change. I didn't observe any VisualGen-quantization-related entries in `tensorrt_llm/usage/llm_args_golden_manifest.json`, either.

To make it 100% safe that the previous failure is only caused by absence of rebase, I'll re-trigger the test to see if `[A10-PyTorch-2]` passes.

**Comment by xrq-phys** [comment]
@zhenhuaw-me 

Hey. I've found the reason that caused skipSm's behavior mismatch against trtllmGen kernels.

We need to fix a default value related to a CuTeDSL-specific optimization that allows skipping `row_max` update when the new `row_max` is within `8 * old_row_max`. This `8.0` rescale threshold is coupling with skipSm affecting its behavior. To resolve, we need to reset this threshold to `1.0`.

This fix is not kernel-side but unfortunately the value is set internally by host side of the kernel module. I'll raise a MR to our internal variant as well, but for TRTLLM here, please allow for one more CI loop as I'm about to push new changes.

Thanks

**Comment by xrq-phys** [comment]
/bot skip --comment "Error not related to VisualGen changes. Multi-GPU tests have all previously succeeded."

**Comment by xrq-phys** [comment]
@zhenhuaw-me I'm feeling confident about merging this now.
